File: go/problem_11.go
package main

import (
	"fmt"
	"sort"
)

// autocomplete returns all strings in words that have prefix s.
// Uses a trie for O(|s| + k) lookup where k is the number of results.
type trieNode struct {
	children map[rune]*trieNode
	isEnd    bool
}

func newTrieNode() *trieNode {
	return &trieNode{children: make(map[rune]*trieNode)}
}

type trie struct {
	root *trieNode
}

func newTrie() *trie { return &trie{root: newTrieNode()} }

// insert adds a word into the trie.
func (t *trie) insert(word string) {
	node := t.root
	for _, ch := range word {
		if node.children[ch] == nil {
			node.children[ch] = newTrieNode()
		}
		node = node.children[ch]
	}
	node.isEnd = true
}

// collect gathers all words reachable from node, prepending prefix.
func collect(node *trieNode, prefix string, results *[]string) {
	if node.isEnd {
		*results = append(*results, prefix)
	}
	for ch, child := range node.children {
		collect(child, prefix+string(ch), results)
	}
}

// autocomplete returns all words sharing the given prefix, in sorted order.
func (t *trie) autocomplete(prefix string) []string {
	node := t.root
	for _, ch := range prefix {
		if node.children[ch] == nil {
			return nil // prefix not found
		}
		node = node.children[ch]
	}
	var results []string
	collect(node, prefix, &results)
	sort.Strings(results)
	return results
}

func problem11() {
	fmt.Println("Problem 11:")
	t := newTrie()
	for _, w := range []string{"dog", "deer", "deal"} {
		t.insert(w)
	}
	got := t.autocomplete("de")
	if len(got) != 2 || got[0] != "deal" || got[1] != "deer" {
		panic("assertion failed")
	}
	if len(t.autocomplete("xyz")) != 0 {
		panic("assertion failed")
	}
	fmt.Println("All Tests Passed")
}

File: python/problem_11.py
class TrieNode:
    def __init__(self):
        self.children: dict[str, 'TrieNode'] = {}
        self.is_end = False


class Trie:
    # Trie supports O(|prefix|) insert and O(|prefix| + k) autocomplete
    # where k is the number of matching words.
    def __init__(self):
        self.root = TrieNode()

    def insert(self, word: str) -> None:
        node = self.root
        for ch in word:
            node = node.children.setdefault(ch, TrieNode())
        node.is_end = True

    def _collect(self, node: TrieNode, prefix: str) -> list[str]:
        results = []
        if node.is_end:
            results.append(prefix)
        for ch, child in sorted(node.children.items()):
            results.extend(self._collect(child, prefix + ch))
        return results

    def autocomplete(self, prefix: str) -> list[str]:
        # returns all words sharing the given prefix, in sorted order
        node = self.root
        for ch in prefix:
            if ch not in node.children:
                return []  # prefix not found
            node = node.children[ch]
        return self._collect(node, prefix)


def main():
    print("Problem 11:")
    t = Trie()
    for w in ["dog", "deer", "deal"]:
        t.insert(w)
    assert t.autocomplete("de") == ["deal", "deer"], "assertion failed"
    assert t.autocomplete("xyz") == [],              "assertion failed"
    print("All Tests Passed")
